src/miral/basic_window_manager.cpp

```diff
@@ -2270,40 +2371,54 @@ void miral::BasicWindowManager::update_windows_for_outputs()
     if (outputs.size() == 0)
         return;
 
-    auto const display_area = outputs.bounding_rectangle();
-
-    for (auto const& window : maximized_surfaces)
+    for (auto& area : window_areas)
     {
-        if (window)
+        for (auto const& window : area->attached_windows)
         {
-            auto& info1 = info_for(window);
+            if (!window)
+                continue;
 
-            Rectangle rect{window.top_left(), window.size()};
+            auto& window_info = info_for(window);
+            Rectangle const original_window_rect{window.top_left(), window.size()};
+            Rectangle updated_window_rect{original_window_rect};
+            Rectangle application_zone = area->application_zone.extents();
+            bool update_window = false;
 
-            switch (info1.state())
+            switch (window_info.state())
             {
             case mir_window_state_maximized:
-                rect = policy->confirm_placement_on_display(info1, mir_window_state_maximized, display_area);
-                place_and_size(info1, rect.top_left, rect.size);
+                updated_window_rect = application_zone;
+                update_window = true;
                 break;
 
             case mir_window_state_horizmaximized:
-                rect.top_left.x = display_area.top_left.x;
-                rect.size.width = display_area.size.width;
-                rect = policy->confirm_placement_on_display(info1, mir_window_state_horizmaximized, rect);
-                place_and_size(info1, rect.top_left, rect.size);
+                updated_window_rect.top_left.x = application_zone.top_left.x;
+                updated_window_rect.size.width = application_zone.size.width;
+                update_window = true;
                 break;
 
             case mir_window_state_vertmaximized:
-                rect.top_left.y = display_area.top_left.y;
-                rect.size.height = display_area.size.height;
-                rect = policy->confirm_placement_on_display(info1, mir_window_state_vertmaximized, rect);
-                place_and_size(info1, rect.top_left, rect.size);
+                updated_window_rect.top_left.y = application_zone.top_left.y;
+                updated_window_rect.size.height = application_zone.size.height;
+                update_window = true;
                 break;
 
             default:
                 break;
             }
+
+            // TODO: Maybe remove update_window and update only if the rect has changed?
+            // This would probably be more correct, but also may or may not break existing users
+            // if (updated_window_rect != original_window_rect)
```

Comment:
1. The state of the windows is not changing. The state of the outputs/application zones is changing, but the policy is notified of that separately.
2. Nowhere is it documented that `confirm_placement_on_display()` will be called on output/application zone changes that don't effect the window
3. Policies can (and should) update window positions in `advise_end()` instead of `confirm_placement_on_display()`, else they may not have enough information to do it properly. For example, the order of `advise_application_zone_delete()`, `advise_output_delete()` and `confirm_placement_on_display()` is not defined, so no particular order should be assumed. The only correct way to remember all useful calls and process them in `advise_end()`